File: tps/gabarito/tp01/TP01-Arquivo-C/arquivo.c

/**
 * @author: Thiago de Campos Ribeiro Nolasco
*/


// Library used:
#include <stdio.h>

// Prototipacao:
void readInput(int quantity);

// receives an input from keyboard and starts the program
int main(void){
	// data declaration
	int value = 0, temp = 0;
	char breakLine = '\n';
	double entries = 0.00;
	
	// Opening the FILE
	FILE *p = fopen("test.txt", "wb+");
	if(p != NULL) {
		scanf("%d", &value);
		temp = value;
		while(temp != 0){
			scanf("%lf", &entries);

			fwrite(&entries, sizeof(double), 1, p);
			fwrite(&breakLine, sizeof(char), 1, p);

			temp--;
		}
	}else{
		printf("Error on opening the File...");
		return -1;
	}
	// Closing FILE
	fclose(p);


	readInput(value);
	// return with no errors
	return 0;
} // ending main

// receives a int "quantity" to read an archive with size of "quantity"
void readInput(int quantity){
	// data declaration
	int foo = 0; 
	double doubleValue = 0.00;
	FILE *p = fopen("test.txt", "rb+");
	
	fseek(p, 0, SEEK_SET);
	
	// Looping that will basicaly:
	// -> Put the file pointer into the beginning of the line (backwards)
	// -> Read a double from the archive, and then, transforms it into integer (narrowing conversion)
	// -> Put the file pointer back into the beginning of the line
	// -> Reads the same double (but without conversion)
	// -> And then, do the comparison...
	for(int i = quantity; i > 0; i--){
		fseek(p, ((sizeof(double) + sizeof(char)) * (i - 1)), SEEK_SET);
		fread(&doubleValue, sizeof(double), 1, p);
		foo = (int) doubleValue;


		if(foo == doubleValue){
			printf("%d\n", foo);
			fseek(p, ((sizeof(double) + sizeof(char)) * (i - 4)) + sizeof(int), SEEK_SET);
		}else{
			// printing only the necessary amount of decimals
			printf("%g\n", doubleValue);
		}		
	}
	// Closing FILE
	fclose(p);
} // ending method

File: tps/gabarito/tp01/TP01-Arquivo-JAVA/Arquivo.java

/**
 * @author: Thiago de Campos Ribeiro Nolasco
*/

// Library used:
import java.io.*;

class Arquivo{
	
	/*
	 * @method: receives a int "quantity" to read an archive with size of "quantity"
	 * @param: quantity Int
	*/
	public static void readInput(int quantity)throws Exception {
		// data declaration
		RandomAccessFile raf = new RandomAccessFile("test.txt", "r");
		int transformedInteger;
		double readDouble, temp;

		// moving the File pointer to the initial line (just fo prevent errors)
		raf.seek(0);

		// Looping that will basicaly:
		// -> Put the file pointer into the beginning of the line (backwards)
		// -> Read a double from the archive, and then, transforms it into integer (narrowing conversion)
		// -> Put the file pointer back into the beginning of the line 
		// -> Reads the same double (but without conversion)
		// -> And then do the comparison...
		for(int i = quantity; i > 0; i--){
			raf.seek((9 * (i - 1)));
			temp = raf.readDouble();
			transformedInteger = (int)temp;
			raf.seek((9 * (i - 1)));
			readDouble = raf.readDouble();

			if(transformedInteger == readDouble){
				System.out.println(transformedInteger);
				raf.seek((9 * (i - 1 )) + 4);
			}else{
				System.out.println(readDouble);
			}
			raf.readByte();
		}
		raf.close();
	}

	/*
	 * @method: (MAIN) receives input from keyboard and starts the program
	 * @param: args String[]
	*/
	public static void main(String[] args)throws Exception {
		// data declaration
		RandomAccessFile raf = new RandomAccessFile("test.txt", "rw");

		int value = 0, temp = 0;
		double entries = 0.00;
		
		value = MyIO.readInt();
		temp = value;
		while(temp != 0){
			entries = MyIO.readDouble();
	
			raf.writeDouble(entries);
			raf.writeBytes("\n");
			temp--;
		}
		raf.close();		

		readInput(value);
	}// ending main
}// ending class
